app/queries/keywords_query.rb

```diff
@@ -1,19 +1,58 @@
 # frozen_string_literal: true
 
 class KeywordsQuery
+  QUERY_PARAMS = [:url_pattern, :keyword_pattern, { link_types: [] }].freeze
+
+  attr_reader :keywords, :url_match_count
+
   def initialize(user)
+    @user = user
     @keywords = user.keywords.select(Keyword.column_names.excluding('html'))
   end
 
-  def call
+  def call(params = {})
+    @keyword_pattern = params[:keyword_pattern]
+    @url_pattern = params[:url_pattern]
+    @link_types = params[:link_types]
+
+    where_name
```

Comment:
Overall the query pattern in this file seems a bit weird, can you refer the mia project for some references https://github.com/nimblehq/mia-web/tree/development/engines. IMO, we can go with some standard naming like

```
   def initalize(relation: Keywords)
      @relation = Keyword.select(Keyword.column_names.excluding('html'))
   end

   def call
       scope = relation
       scope = filter_by_name(scope) if keyword_pattern.present?
       scope = filter_by_result_links(scope) if filter_on_result_link?
   end

```

Reply:
This would improve the overall readability, in current code, the instance variable is changed anywhere which makes it difficult to track

Reply:
Very interesting, thank you @ankitkalia1195 

I'm trying to implement it, but I have 2 concerns:
1. I need to return 2x things, the keyword list as well as the url_match_count. What would you recommend?
    - return `self` as currently and access these 2 values from public attr_readers?
    - return an array of both values (it will make the controller code somehow less clean)
    - split into 2x different query objects, one for each value. To keep it DRY, I might need a module to include shared code
2. The `scope` example solve 1 issue (keep a non filtered relation for sub-queries) but it feel like we moved from OOP to a more procedural approach, is that an issue for the query object pattern?

Thanks a lot! @junan  feel free to share your insights too :)

Reply:
@ankitkalia1195 @junan I have just pushed a first tentative. Feel free to comment it :) 
Thanks!

Reply:
I checked the MIA codebase and I like this [RevenueQuery](https://github.com/nimblehq/mia-web/blob/development/engines/mia_admin/app/queries/mia_admin/revenue_reports/revenue_query.rb) implementation. It implemented a separate public method for each required data which is more readable/testable and validates **Single-responsibility principle** for this case(where we need multiple values).

In the current implementation, we are returning the value for a single method(returning `keywords` and `url_match_count` for `call` ) which violates the ingle-responsibility principle.

In my opinion [RevenueQuery](https://github.com/nimblehq/mia-web/blob/development/engines/mia_admin/app/queries/mia_admin/revenue_reports/revenue_query.rb) implementation approach is better in this case.(We could access these 2 values from public attr_readers also but I find the RevenueQuery implementation is more reliable and testable)

@malparty @ankitkalia1195  What do you think?

Reply:
Nice!! I did not find this one ☝️  
I totally validate that this will work fine here, thanks @junan. 

I did not know if it was acceptable to have many public methods instead of the generic call for a query object pattern. If it is OK, then hell yeah! :) 

I'll implement it in a few minutes ;-)

Reply:
@ankitkalia1195 @junan code updated :)